### PR TITLE
Allow Union[unicode, str] rather than just str in several places

### PR DESCRIPTION
**Pull request by bsolomon1124**
Closses https://github.com/python/typeshed/issues/3844.

Appreciate a review of this because I ended up using not just `Union[str, unicode]` but also `TypeVar('_AnyText', str, bytes, unicode)` in some places, specifically for functions that accept any of str/bytes/unicode but then produce a result of type that is dependent upon the input type.

Conversely, some functions are not that deterministic, so in those places I stuck with just Union; for instance, `urlunparse()` could (theoretically) take a `ParseResult` input that is mixed unicode and str:

```python
>>> urlunparse(ParseResult(scheme='http', netloc='www.google.com', path=str('/me'), params=str(''), query='foo=bar', fragment=''))
u'http://www.google.com/me?foo=bar'
```

### CODE REVIEW

File: stdlib/2/urlparse.pyi [hunk 1]
```diff
@@ -1,8 +1,9 @@
 # Stubs for urlparse (Python 2)
 
-from typing import AnyStr, Dict, List, NamedTuple, Tuple, Sequence, Union, overload, Optional
+from typing import AnyStr, Dict, List, NamedTuple, Tuple, TypeVar, Sequence, Union, overload, Optional
 
 _String = Union[str, unicode]
+_AnyText = TypeVar('_AnyText', str, bytes, unicode)
```
Comment (JelleZijlstra) [inline stdlib/2/urlparse.pyi]: This is just `typing.AnyStr` (str and bytes are the same in Python 2).

File: stdlib/2and3/threading.pyi [hunk 2]
```diff
@@ -63,14 +65,14 @@ class Thread:
     else:
         def __init__(self, group: None = ...,
                      target: Optional[Callable[..., Any]] = ...,
-                     name: Optional[str] = ...,
+                     name: Optional[_String] = ...,
```
Comment (JelleZijlstra) [inline stdlib/2and3/threading.pyi]: You can just use `typing.Text` in argument types; it implicitly includes str.

Comment (bsolomon1124) [reply stdlib/2and3/threading.pyi]: @JelleZijlstra Okay - I noticed that `_String = Union[str, unicode]` was defined in `urlparse.pyi` (prior to this PR); is there a reason for that?

Comment (JelleZijlstra) [reply stdlib/2and3/threading.pyi]: In return types it may make a bit more of a difference, or it may just be unnecessary.

Comment (bsolomon1124) [reply stdlib/2and3/threading.pyi]: > In return types it may make a bit more of a difference

Can you clarify what you mean by this?

Comment (srittau) [reply stdlib/2and3/threading.pyi]: `Text` is equivalent to `Union[str, unicode]` in argument types, but in other places (like return types) it is equivalent to just `unicode`. (For Python 2.)

Comment (bsolomon1124) [reply stdlib/2and3/threading.pyi]: Thanks for the clarification, corrected in 7ba0cc02.

File: stdlib/2/urlparse.pyi (hunk 1 above)
Comment (bsolomon1124) [reply stdlib/2/urlparse.pyi]: Corrected in 003dd638.

**Comment by gvanrossum** [comment]
Note that some of these were rolled back in #4437.

**Pull request merged**